### PR TITLE
Print NFA to mata format

### LINKED ISSUE #168: true/false in mata format
Right now in the format there are two ways to write true/false: either `\true` and `\false` or `true` and `false`. We should decide which one we want to keep, right now we use the second one (`true` and `false`).
Also, why is there still `FORMAT.md` file?

### LINKED ISSUE #202: Print NFA to `.mata` format
We need a function to print NFA to our Mata format, similar to `Nfa::Nfa::print_to_DOT()`.

### PR DESCRIPTION
**Pull request by jurajsic**
This PR adds function `print_to_mata` which prints NFA to mata format. It also updates .mata format a bit.

### CODE REVIEW

File: AUTOMATAFORMAT.md
Comment (Adda0) [inline AUTOMATAFORMAT.md]: I suggest changing the tile so that it is immediately visible that this is Mata format, like so:
```suggestion
# The Mata file format for automata (.mata format)
```

File: AUTOMATAFORMAT.md
Comment (Adda0) [inline AUTOMATAFORMAT.md]: Slightly better wording.
```suggestion
* There are **words with a special meaning**: `\true`,`\false`, `\min`, `\max`. The first two represent true/false in Boolean formulas.
```

File: AUTOMATAFORMAT.md
Comment (Adda0) [inline AUTOMATAFORMAT.md]:
```suggestion
* **Character class**: Character classes, i.e. sets of symbols as used in normal regular expressions, are of the form `[bla]` or `[^bla]` where `bla` is a sequence of symbols and intervals of symbols such as `a-z`, for instance `abcd-hij-z` denotes all lower case alphabet symbols. The `^` in front complements the entire class. The character `-` in sincluded as `\-`, analogously `\` and `^`. The special keywords `\min` and `\max` denote the first repsective the last letter the alphabet.
```

File: AUTOMATAFORMAT.md
Comment (Adda0) [inline AUTOMATAFORMAT.md]:
```suggestion
* Transition lines are of the form `"<State> <Formula>"` representing the source state and the transition formula - a Boolean formula over symbol literals, states, and nodes.
```

File: AUTOMATAFORMAT.md
Comment (Adda0) [inline AUTOMATAFORMAT.md]:
```suggestion
The transition and alphabet type is encoded in the name of a section, e.g., `@NFA-explicit` denotes that a nondeterministic finite automaton over an explicit alphabet is given.
```

File: include/mata/nfa.hh
Comment (Adda0) [inline include/mata/nfa.hh]: The `@brief` needs an empty line after the brief description. As per [Doxygen documentation](https://www.doxygen.nl/manual/commands.html#cmdbrief).
```suggestion
     * @brief Prints the automaton in mata format
     *
     * If you need to parse the automaton again, use IntAlphabet in construct()
```

File: include/mata/nfa.hh
Comment (Adda0) [inline include/mata/nfa.hh]: The `@brief` needs an empty line after the brief description. As per [Doxygen documentation](https://www.doxygen.nl/manual/commands.html#cmdbrief).

```suggestion
     * @brief Prints the automaton to the output stream in mata format
     *
     * If you need to parse the automaton again, use IntAlphabet in construct()
```

File: src/nfa/nfa.cc
Comment (Adda0) [inline src/nfa/nfa.cc]: Is seems strange to use `tran` (as in, transition) if the value is a move (as in, a symbol and a set of target states). We can keep it here, but I would encourage you to use `move` if possible in such cases. It is more readable, I think.

File: src/nfa/nfa.cc
Comment (Adda0) [inline src/nfa/nfa.cc:1330]: What do you need for this? Could we modify alphabets to be able to do so? I will have to check what `@Alphabet-numbers` is. Is `@Alphabet-numbers` even described in AUTOMATAFORMAT.md? I cannot find it there. Maybe we should have the alphabet names used in our format beside its descriptions.

File: tests/nfa/nfa.cc
Comment (Adda0) [inline tests/nfa/nfa.cc]: A beautiful one-liner, indeed. I would split the line into multiple smaller parts, so that it is more readable, however. After all, this should serve as an example people will look at, too.

**Review by Adda0 — APPROVED** [review]
Looks great to me. That is one more important and much needed feature implemented.

File: src/nfa/nfa.cc
Comment (jurajsic) [reply src/nfa/nfa.cc:1330]: It is mentioned under `Distinguishing between states, symbols, and nodes`. For it to work, we would have to modify the parser, currently it does not support it at all. This is why I need to use `IntAlphabet` if I want to parse output of this function, so that it creates the same symbols.

File: src/nfa/nfa.cc
Comment (jurajsic) [reply src/nfa/nfa.cc]: Fixed

File: tests/nfa/nfa.cc
Comment (jurajsic) [reply tests/nfa/nfa.cc]: Done

File: include/mata/nfa.hh
Comment (tfiedor) [inline include/mata/nfa.hh:611]: Maybe, we could unify this with name of parsing function? I believe it is `parse_mf`, so maybe it could be `print_mf`/`dump_mf`/`write_mf` or something similar?

File: tests/nfa/nfa.cc
Comment (tfiedor) [inline tests/nfa/nfa.cc:2987]: For harder test,  you could traverse all files in https://github.com/VeriFIT/mata/tree/devel/tests-performance/automata directory and do parse/print and comparison as well.

**Review by tfiedor — APPROVED** [review]
Seems suspiciously easy. I would consider trying to convert those in `tests-performace\automata` folder, as it might show some bugs/cases you did not consider. Also consider adding some throwing of exceptions, e.g., when some alphabets are not supported.

File: src/nfa/nfa.cc
Comment (Adda0) [reply src/nfa/nfa.cc:1330]: Got it. Thank you for the explanation. Seems good to me for now, then.

File: include/mata/nfa.hh
Comment (jurajsic) [reply include/mata/nfa.hh:611]: Yeah, we should probably unify the names, but first we need to decide which one to use. Is `_mf` all right? Should we keep that?

Comment (tfiedor) [reply include/mata/nfa.hh:611]: Well, now that you are saying maybe `_mata` is better, since the files have `.mata` extension :D Or change or files to `.mf`

**Comment by jurajsic** [comment]
> Seems suspiciously easy. I would consider trying to convert those in `tests-performace\automata` folder, as it might show some bugs/cases you did not consider. Also consider adding some throwing of exceptions, e.g., when some alphabets are not supported.

I currently do not take alphabets into any consideration for printing, I just print the exact value of the symbol. I think we need to sort out alphabets and how to work with them before implementing this. But I will add a TODO.

File: include/mata/nfa.hh
Comment (jurajsic) [reply include/mata/nfa.hh:611]: Then maybe we should decide how the mata file format extension should look first. I'm for `.mata`.

Comment (Adda0) [reply include/mata/nfa.hh:611]: Definitely `_mata`. And I thought we already decided on `.mata` for file extension name. Do we want to discuss this with others first?

Comment (kilohsakul) [reply include/mata/nfa.hh:611]: don't remember, .mata and _mata sounds better to me also.

Comment (tfiedor) [reply include/mata/nfa.hh:611]: Or `.mata_the_ultimate_automata_format_not_to_be_confused_with_maťa`

**Comment by Adda0** [comment]
Conflicts resolved.

Do we want to rename the `parse_mf`, etc. functions in this PR, too?

Otherwise, we can merge the PR.

File: tests/nfa/nfa.cc
Comment (Adda0) [reply tests/nfa/nfa.cc:2987]: We are confident that the introduced changes work as expected, and we do not need to test it further. If we ever find any bugs with the print function, we will rethink our stance on this issue. We want this merged as quickly as possible so that we can start using it for the currently worked-on features.

**Comment by Adda0** [comment]
All discussion points have been resolved. I will merge the PR ~~now~~ after Lukáš checks the new format description.

**Pull request merged**
